### ISSUE TITLE
Systematize the scm methods

### ISSUE THREAD

**Issue by meling** [quickfeed/quickfeed#373]
The scm package can sometimes be useful for other tools (outside the quickfeed ecosystem) as well. Hence it could be made into its own module that can be used without downloading the entire quickfeed module. Maybe it could be made more generic as well, and could have better support for a selection of git hosting services. It should provide sufficient set of tests and examples and maybe provide a command line tool that implements most or all features.

**Comment by meling** [comment]
Here is the list of 26 potential functions from the `web` package to be migrated to `scm` as higher-level tasks:

```shell
% grep func *.go | grep scm.SCM
access_control.go:func (s *AutograderService) getSCM(ctx context.Context, user *pb.User, provider string) (scm.SCM, error) {
access_control.go:func (s *AutograderService) getUserAndSCM(ctx context.Context, provider string) (*pb.User, scm.SCM, error) {
access_control.go:func (s *AutograderService) getUserAndSCMForCourse(ctx context.Context, courseID uint64) (*pb.User, scm.SCM, error) {
access_control.go:func hasTeacherScopes(ctx context.Context, sc scm.SCM) bool {
courses.go:func (s *AutograderService) updateEnrollment(ctx context.Context, sc scm.SCM, curUser string, request *pb.Enrollment) error {
courses.go:func (s *AutograderService) updateEnrollments(ctx context.Context, sc scm.SCM, cid uint64) error {
courses.go:func (s *AutograderService) updateCourse(ctx context.Context, sc scm.SCM, request *pb.Course) error {
courses.go:func (s *AutograderService) rejectEnrollment(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
courses.go:func (s *AutograderService) enrollStudent(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
courses.go:func (s *AutograderService) enrollTeacher(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
courses_new.go:func (s *AutograderService) createCourse(ctx context.Context, sc scm.SCM, request *pb.Course) (*pb.Course, error) {
groups.go:func (s *AutograderService) deleteGroup(ctx context.Context, sc scm.SCM, request *pb.GroupRequest) error {
groups.go:func (s *AutograderService) updateGroup(ctx context.Context, sc scm.SCM, request *pb.Group) error {
organizations.go:func (s *AutograderService) getOrganization(ctx context.Context, sc scm.SCM, org string, user string) (*pb.Organization, error) {
repositories.go:func (s *AutograderService) isEmptyRepo(ctx context.Context, sc scm.SCM, request *pb.RepositoryRequest) error {
scm_helpers.go:func createRepoAndTeam(ctx context.Context, sc scm.SCM, course *pb.Course, group *pb.Group) (*pb.Repository, *scm.Team, error) {
scm_helpers.go:func deleteGroupRepoAndTeam(ctx context.Context, sc scm.SCM, repositoryID, teamID, orgID uint64) error {
scm_helpers.go:func createStudentRepo(ctx context.Context, sc scm.SCM, org *pb.Organization, path string, student string) (*scm.Repository, error) {
scm_helpers.go:func addUserToStudentsTeam(ctx context.Context, sc scm.SCM, organizationPath string, userName string) error {
scm_helpers.go:func promoteUserToTeachersTeam(ctx context.Context, sc scm.SCM, organizationPath string, userName string) error {
scm_helpers.go:func updateReposAndTeams(ctx context.Context, sc scm.SCM, course *pb.Course, login string, state pb.Enrollment_UserStatus) (*scm.Repository, error) {
scm_helpers.go:func grantAccessToCourseRepos(ctx context.Context, sc scm.SCM, org, login string) error {
scm_helpers.go:func updateGroupTeam(ctx context.Context, sc scm.SCM, group *pb.Group, orgID uint64) error {
scm_helpers.go:func removeUserFromCourse(ctx context.Context, sc scm.SCM, login string, repo *pb.Repository) error {
scm_helpers.go:func revokeTeacherStatus(ctx context.Context, sc scm.SCM, org, userName string) error {
scm_helpers.go:func isEmpty(ctx context.Context, sc scm.SCM, repos []*pb.Repository) error {
```

**Comment by meling** [comment]
Here is the list of 26 potential functions from the `web` package to be migrated to `scm` as higher-level tasks:

```shell
% grep func *.go | grep scm.SCM
access_control.go:func (s *AutograderService) getSCM(ctx context.Context, user *pb.User, provider string) (scm.SCM, error) {
access_control.go:func (s *AutograderService) getUserAndSCM(ctx context.Context, provider string) (*pb.User, scm.SCM, error) {
access_control.go:func (s *AutograderService) getUserAndSCMForCourse(ctx context.Context, courseID uint64) (*pb.User, scm.SCM, error) {
access_control.go:func hasTeacherScopes(ctx context.Context, sc scm.SCM) bool {
courses.go:func (s *AutograderService) updateEnrollment(ctx context.Context, sc scm.SCM, curUser string, request *pb.Enrollment) error {
courses.go:func (s *AutograderService) updateEnrollments(ctx context.Context, sc scm.SCM, cid uint64) error {
courses.go:func (s *AutograderService) updateCourse(ctx context.Context, sc scm.SCM, request *pb.Course) error {
courses.go:func (s *AutograderService) rejectEnrollment(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
courses.go:func (s *AutograderService) enrollStudent(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
courses.go:func (s *AutograderService) enrollTeacher(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
courses_new.go:func (s *AutograderService) createCourse(ctx context.Context, sc scm.SCM, request *pb.Course) (*pb.Course, error) {
groups.go:func (s *AutograderService) deleteGroup(ctx context.Context, sc scm.SCM, request *pb.GroupRequest) error {
groups.go:func (s *AutograderService) updateGroup(ctx context.Context, sc scm.SCM, request *pb.Group) error {
organizations.go:func (s *AutograderService) getOrganization(ctx context.Context, sc scm.SCM, org string, user string) (*pb.Organization, error) {
repositories.go:func (s *AutograderService) isEmptyRepo(ctx context.Context, sc scm.SCM, request *pb.RepositoryRequest) error {
scm_helpers.go:func createRepoAndTeam(ctx context.Context, sc scm.SCM, course *pb.Course, group *pb.Group) (*pb.Repository, *scm.Team, error) {
scm_helpers.go:func deleteGroupRepoAndTeam(ctx context.Context, sc scm.SCM, repositoryID, teamID, orgID uint64) error {
scm_helpers.go:func createStudentRepo(ctx context.Context, sc scm.SCM, org *pb.Organization, path string, student string) (*scm.Repository, error) {
scm_helpers.go:func addUserToStudentsTeam(ctx context.Context, sc scm.SCM, organizationPath string, userName string) error {
scm_helpers.go:func promoteUserToTeachersTeam(ctx context.Context, sc scm.SCM, organizationPath string, userName string) error {
scm_helpers.go:func updateReposAndTeams(ctx context.Context, sc scm.SCM, course *pb.Course, login string, state pb.Enrollment_UserStatus) (*scm.Repository, error) {
scm_helpers.go:func grantAccessToCourseRepos(ctx context.Context, sc scm.SCM, org, login string) error {
scm_helpers.go:func updateGroupTeam(ctx context.Context, sc scm.SCM, group *pb.Group, orgID uint64) error {
scm_helpers.go:func removeUserFromCourse(ctx context.Context, sc scm.SCM, login string, repo *pb.Repository) error {
scm_helpers.go:func revokeTeacherStatus(ctx context.Context, sc scm.SCM, org, userName string) error {
scm_helpers.go:func isEmpty(ctx context.Context, sc scm.SCM, repos []*pb.Repository) error {
```

**Comment by 0xf8f8ff** [comment]
Will be partially solved by #864. However, this will not cover the "separate module" part.
